File: leetcode_problems/go/lc105_binary_tree_from_preorder_and_inorder.go
package main

func buildTreeHelper(preorder []int, preStart, preEnd int, inorder []int, inStart, inEnd int, inMap map[int]int) *TreeNode {
	if preStart > preEnd || inStart > inEnd {
		return nil
	}
	rootVal := preorder[preStart]
	root := &TreeNode{Val: rootVal}
	inRoot := inMap[rootVal]
	leftSize := inRoot - inStart

	root.Left = buildTreeHelper(preorder, preStart+1, preStart+leftSize, inorder, inStart, inRoot-1, inMap)
	root.Right = buildTreeHelper(preorder, preStart+leftSize+1, preEnd, inorder, inRoot+1, inEnd, inMap)
	return root
}

func buildTreeDivideAndConquer(preorder []int, inorder []int) *TreeNode {
	inMap := make(map[int]int)
	for i, v := range inorder {
		inMap[v] = i
	}
	return buildTreeHelper(preorder, 0, len(preorder)-1, inorder, 0, len(inorder)-1, inMap)
}

File: leetcode_problems/python/lc105_binary_tree_from_preorder_and_inorder.py
from utils import TreeNode


class Solution:
    def buildTree(self, preorder: list[int], inorder: list[int]) -> TreeNode | None:
        def build(preorder: list[int], inorder: list[int]) -> TreeNode | None:
            if not preorder or not inorder:
                return None

            rootVal = preorder.pop(0)
            inPos = inorder.index(rootVal)
            root = TreeNode(rootVal)
            root.left = build(preorder, inorder[:inPos])
            root.right = build(preorder, inorder[inPos + 1 :])
            return root

        # Оптимизация с маппером не работает, т.к. мы режем массив
        # На каждом шаге рекурсии у нас меняются индексы массива
        return build(preorder, inorder)
